## Data Preprocessing

This notebook illustrates steps required to preprocess Wyscout football match event data. Specifically, data is extracted and preprocessed for leagues in Italy, England, Spain, France and Germany.

### Libaries

In [26]:
# import required libraries
import os
import ast
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from mplsoccer import Pitch, VerticalPitch

### Import Wyscout Public Match Event Dataset

Wyscout Public Match Event Dataset is available as `PublicWyscoutLoader` object in `socceraction` library.

In [24]:
# import wyscout public match event data loader from socceraction library
from socceraction.data.wyscout import PublicWyscoutLoader 

# load public wyscout data
wyscout_data = PublicWyscoutLoader()

In [28]:
# view all available competitions
wyscout_data.competitions()

competition_id  season_id   country_name        competition_name  \
0             524     181248          Italy  Italian first division   
1             364     181150        England  English first division   
2             795     181144          Spain  Spanish first division   
3             412     181189         France   French first division   
4             426     181137        Germany   German first division   
5             102       9291  International   European Championship   
6              28      10078  International               World Cup   

  competition_gender season_name  
0               male   2017/2018  
1               male   2017/2018  
2               male   2017/2018  
3               male   2017/2018  
4               male   2017/2018  
5               male        2016  
6               male        2018

Retrieve event data for all matches in Italy, England, Spain, France and Germany.

### EPL Data Preprocessing

#### Download Wyscout data for EPL

Wyscout event data for each match is downloaded and saved as .csv file. In addition, all .csv files are saved into a single `events.csv` file. 

In [ ]:
# England 17/18, competition_id = 364, season_id = 181150
#epl_games = wyscout_data.games(competition_id = 364, season_id = 181150)["game_id"]

##################################################################################################
# Below section were executed a single time and saved as .csv files (only comment out if needed)#
##################################################################################################
# download all premier league matches and save as .csv files
#events = pd.DataFrame()
#for i in epl_games:
#    df = wyscout_data.events(i)
#    events = pd.concat([events, df])

# remove first columns with indices since I forgot to add index = False when saving .csv files
#events = events.drop('Unnamed: 0', axis = 1)
#events.to_csv(f'data/epl/events.csv', index = False) 

In [38]:
events.head()

event_id  game_id  period_id  milliseconds  team_id  player_id  type_id  \
0  177959171  2499719          1      2758.649     1609      25413        8   
1  177959172  2499719          1      4946.850     1609     370224        8   
2  177959173  2499719          1      6542.188     1609       3319        8   
3  177959174  2499719          1      8143.395     1609     120339        8   
4  177959175  2499719          1     10302.366     1609     167145        8   

  type_name  subtype_id subtype_name  \
0      Pass          85  Simple pass   
1      Pass          83    High pass   
2      Pass          82    Head pass   
3      Pass          82    Head pass   
4      Pass          85  Simple pass   

                                  positions            tags  
0  [{'y': 49, 'x': 49}, {'y': 78, 'x': 31}]  [{'id': 1801}]  
1  [{'y': 78, 'x': 31}, {'y': 75, 'x': 51}]  [{'id': 1801}]  
2  [{'y': 75, 'x': 51}, {'y': 71, 'x': 35}]  [{'id': 1801}]  
3  [{'y': 71, 'x': 35}, {'y': 95, 'x': 41}]  [{'id': 1801}]  
4  [{'y': 95, 'x': 41}, {'y': 88, 'x': 72}]  [{'id': 1801}]

In [39]:
events.shape

(643150, 12)

There are in total 643150 events in 380 EPL matches.

#### Desirable formatting of raw events data

In [41]:
# upload 'events' dataframe that includes all events of all 380 EPL games
df = pd.read_csv('data/epl/events.csv')

In [43]:
# view columns
df.columns

Index(['event_id', 'game_id', 'period_id', 'milliseconds', 'team_id',
       'player_id', 'type_id', 'type_name', 'subtype_id', 'subtype_name',
       'positions', 'tags'],
      dtype='object')

In [46]:
# remove columns 'event id' and 'milliseconds'
rm_col_ind = np.r_[0, 3]
df = df.drop(columns = df.columns[rm_col_ind])

In [48]:
df.head()

game_id  period_id  team_id  player_id  type_id type_name  subtype_id  \
0  2499719          1     1609      25413        8      Pass          85   
1  2499719          1     1609     370224        8      Pass          83   
2  2499719          1     1609       3319        8      Pass          82   
3  2499719          1     1609     120339        8      Pass          82   
4  2499719          1     1609     167145        8      Pass          85   

  subtype_name                                 positions            tags  
0  Simple pass  [{'y': 49, 'x': 49}, {'y': 78, 'x': 31}]  [{'id': 1801}]  
1    High pass  [{'y': 78, 'x': 31}, {'y': 75, 'x': 51}]  [{'id': 1801}]  
2    Head pass  [{'y': 75, 'x': 51}, {'y': 71, 'x': 35}]  [{'id': 1801}]  
3    Head pass  [{'y': 71, 'x': 35}, {'y': 95, 'x': 41}]  [{'id': 1801}]  
4  Simple pass  [{'y': 95, 'x': 41}, {'y': 88, 'x': 72}]  [{'id': 1801}]

Columns `positions` and `tags` are in a list format but when saved as `.csv` file they are converted into a string format. To work with these columns `literal_eval` function from `ast` library is used.

In [49]:
# convert strings into python lists
df['tags'] = df['tags'].apply(ast.literal_eval)
df['positions'] = df['positions'].apply(ast.literal_eval)

# make 'type_name' and 'subtype_name' columns lowercase 
df['type_name'] = df['type_name'].str.lower()
df['subtype_name'] = df['subtype_name'].str.lower()

# create separate initial(start) and final(end) coordinates from 'positions' column
# if action has only 'start' coordinates set 'end' coordinates to 'nan'
df['x_start'] = df['positions'].apply(lambda x: x[0]['x'])
df['y_start'] = df['positions'].apply(lambda x: x[0]['y'])
df['x_end'] = df['positions'].apply(lambda x: x[1]['x'] if len(x) == 2 else np.nan)
df['y_end'] = df['positions'].apply(lambda x: x[1]['y'] if len(x) == 2  else np.nan)

In [51]:
df.head()

game_id  period_id  team_id  player_id  type_id type_name  subtype_id  \
0  2499719          1     1609      25413        8      pass          85   
1  2499719          1     1609     370224        8      pass          83   
2  2499719          1     1609       3319        8      pass          82   
3  2499719          1     1609     120339        8      pass          82   
4  2499719          1     1609     167145        8      pass          85   

  subtype_name                                 positions            tags  \
0  simple pass  [{'y': 49, 'x': 49}, {'y': 78, 'x': 31}]  [{'id': 1801}]   
1    high pass  [{'y': 78, 'x': 31}, {'y': 75, 'x': 51}]  [{'id': 1801}]   
2    head pass  [{'y': 75, 'x': 51}, {'y': 71, 'x': 35}]  [{'id': 1801}]   
3    head pass  [{'y': 71, 'x': 35}, {'y': 95, 'x': 41}]  [{'id': 1801}]   
4  simple pass  [{'y': 95, 'x': 41}, {'y': 88, 'x': 72}]  [{'id': 1801}]   

   x_start  y_start  x_end  y_end  
0       49       49   31.0   78.0  
1       31       78   51.0   75.0  
2       51       75   35.0   71.0  
3       35       71   41.0   95.0  
4       41       95   72.0   88.0

To map tag ids in `tags` column to their definitions I use a separate `wyscout_tags.csv` file. This file is not a part of `PublicWyscoutLoader`, and it can be downloaded from [this link](https://figshare.com/articles/dataset/Mapping_of_tag_identifiers_to_tag_names/11743818?backTo=/collections/Soccer_match_event_dataset/4415000).

In [52]:
# import tags for encode subevents in event data
tags = pd.read_csv('data/wyscout_tags.csv', sep = ';')

# make all descriptions lowercase
tags['Description'] = tags['Description'].str.lower()

# transform tags data frame into dictionary
tags = dict(zip(tags['Tag'], tags['Description']))

# use dictionaries and list comprehensions to convert tags into tag ids and their descriptions 
df['tag_id'] = df['tags'].apply(lambda x: [value for d in x for value in d.values()])
df['tag_name'] = df['tag_id'].apply(lambda x: [tags[i] for i in x])

# drop redundant column 'positions'
df.drop(columns = ['positions', 'tags'], inplace=True)

In [53]:
df.head()

game_id  period_id  team_id  player_id  type_id type_name  subtype_id  \
0  2499719          1     1609      25413        8      pass          85   
1  2499719          1     1609     370224        8      pass          83   
2  2499719          1     1609       3319        8      pass          82   
3  2499719          1     1609     120339        8      pass          82   
4  2499719          1     1609     167145        8      pass          85   

  subtype_name  x_start  y_start  x_end  y_end  tag_id    tag_name  
0  simple pass       49       49   31.0   78.0  [1801]  [accurate]  
1    high pass       31       78   51.0   75.0  [1801]  [accurate]  
2    head pass       51       75   35.0   71.0  [1801]  [accurate]  
3    head pass       35       71   41.0   95.0  [1801]  [accurate]  
4  simple pass       41       95   72.0   88.0  [1801]  [accurate]

For each event, there can be a multiple number of tags. Thus, these tags are stored as lists.

In [58]:
# rearrange columns using only required indices and passing them into np.r_
rearr_cols = np.r_[0:8, 12, 13, 8:12]
df = df.iloc[:, rearr_cols]

In [59]:
df.head()

game_id  period_id  team_id  player_id  type_id type_name  subtype_id  \
0  2499719          1     1609      25413        8      pass          85   
1  2499719          1     1609     370224        8      pass          83   
2  2499719          1     1609       3319        8      pass          82   
3  2499719          1     1609     120339        8      pass          82   
4  2499719          1     1609     167145        8      pass          85   

  subtype_name  tag_id    tag_name  x_start  y_start  x_end  y_end  
0  simple pass  [1801]  [accurate]       49       49   31.0   78.0  
1    high pass  [1801]  [accurate]       31       78   51.0   75.0  
2    head pass  [1801]  [accurate]       51       75   35.0   71.0  
3    head pass  [1801]  [accurate]       35       71   41.0   95.0  
4  simple pass  [1801]  [accurate]       41       95   72.0   88.0

In [60]:
# save 'df' as 'refined_events.csv' dataframe
df.to_csv('data/epl/refined_events.csv', index = False)    

In [61]:
df.shape

(643150, 14)

#### Extract EPL shot event data 

In [64]:
# free kicks are not included (penalties are also part of free kicks)
shots = df[df['type_name'] == 'shot']

In [119]:
shots.head()

game_id  period_id  team_id  player_id  type_id type_name  subtype_id  \
46   2499719          1     1609      25413       10      shot         100   
62   2499719          1     1631      26150       10      shot         100   
91   2499719          1     1631      14763       10      shot         100   
128  2499719          1     1609       7868       10      shot         100   
249  2499719          1     1609       7868       10      shot         100   

    subtype_name                       tag_id  \
46          shot  [101, 402, 201, 1205, 1801]   
62          shot       [401, 201, 1211, 1802]   
91          shot  [101, 403, 201, 1207, 1801]   
128         shot       [401, 201, 1215, 1802]   
249         shot       [402, 201, 1205, 1801]   

                                              tag_name  x_start  y_start  \
46   ['goal', 'right foot', 'opportunity', 'positio...       88       41   
62   ['left foot', 'opportunity', 'position: out ce...       85       52   
91   ['goal', 'head/body', 'opportunity', 'position...       96       52   
128  ['left foot', 'opportunity', 'position: out hi...       81       33   
249  ['right foot', 'opportunity', 'position: goal ...       75       30   

     x_end  y_end  
46     0.0    0.0  
62   100.0  100.0  
91   100.0  100.0  
128    0.0    0.0  
249    0.0    0.0

In [120]:
# save 'df' as 'refined_events.csv' dataframe
shots.to_csv('data/epl/shots.csv', index = False)    

### LaLiga Data Preprocessing

Since data retrieval and preprocessing steps are similar to what was data for EPL data, all the above code is implemented in a single cell.

In [ ]:
# Spain 17/18, competition_id = 795, season_id = 181144
laliga_games = wyscout_data.games(competition_id = 795, season_id = 181144)["game_id"]

#events = pd.DataFrame()
#for i in laliga_games:
#    print(i)
#    df = wyscout_data.events(i)
#    df.to_csv(f'data/laliga/laliga_games/{i}.csv')
#    events = pd.concat([events, df])

# remove first columns with indices since I forgot to add index = False when saving .csv files
#events = events.drop('Unnamed: 0', axis = 1)
#events.to_csv(f'data/laliga/events.csv', index = False) 